server/src-lib/Hasura/Server/Auth/JWT.hs

```diff
@@ -0,0 +1,225 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+
+module Hasura.Server.Auth.JWT
+  ( processJwt
+  , RawJWT
+  , JWTConfig (..)
+  ) where
+
+import           Control.Lens
+import           Control.Monad              (when)
+import           Crypto.JOSE.Types          (Base64Integer (..))
+import           Crypto.JWT
+import           Crypto.PubKey.RSA          (PublicKey (..))
+import           Data.ASN1.BinaryEncoding   (DER (..))
+import           Data.ASN1.Encoding         (decodeASN1')
+import           Data.ASN1.Types            (ASN1 (End, IntVal, Start),
+                                             ASN1ConstructionType (Sequence),
+                                             fromASN1)
+import           Data.List                  (find)
+import           Data.Time.Clock            (getCurrentTime)
+import           Hasura.Prelude
+import           Hasura.RQL.Types
+import           Hasura.Server.Utils        (userRoleHeader)
+
+import qualified Data.Aeson                 as A
+import qualified Data.ByteString.Lazy       as BL
+import qualified Data.ByteString.Lazy.Char8 as BLC
+import qualified Data.CaseInsensitive       as CI
+import qualified Data.HashMap.Strict        as Map
+import qualified Data.PEM                   as PEM
+import qualified Data.Text                  as T
+import qualified Data.Text.Encoding         as TE
+import qualified Data.X509                  as X509
+import qualified Network.HTTP.Types         as HTTP
+
+
+type RawJWT = BL.ByteString
+
+-- | Process the request headers to verify the JWT and extract UserInfo from it
+processJwt
+  :: ( MonadIO m
+     , MonadError QErr m)
+  => JWK
+  -> HTTP.RequestHeaders
+  -> m UserInfo
+processJwt key headers = do
+
+  -- try to parse JWT token from Authorization header
+  let mAuthzHeader = find (\h -> fst h == CI.mk "Authorization") headers
+  (_, authzHeader) <- maybe missingAuthzHeader return mAuthzHeader
+  let tokenParts = BLC.words $ BL.fromStrict authzHeader
```

Comment:
pattern match to throw errors instead of checking for length and accessing an element with an index